Trivial serialisation plugin mechanism

fwyzard opened the pull request:
#### PR description:

Introduce `ngt::SerialiserFactory`, a plugin-based trivial serialiser, and the serialisation plugins for all `PortableHostCollection`- and `PortableHostObject`-derived data formats.

This relies on some new components:
  * `ngt::AnyBuffer`: `ngt::AnyBuffer` behaves like `std::any`, with two differences: it can only be used with trivially copyable types, and provides access to the underlying memory buffer to allow memcpy'ing its content.
 * the `ngt::MemoryCopyTraits<T>` type traits. The struct `ngt::MemoryCopyTraits<T>` should be specialised for the types that can be safely `memcpy`'ed.

And is accompanied by various tests:
  * Catch2 tests for the `ngt::MemoryCopyTraits`.
  * Catch2 tests for the `ngt::SerialiserFactory`.
  * integration tasts based on the `ngt::GenericCloner`.

The `ngt::GenericCloner` is included to demonstrate how the new functionalities can be used, and to implement various integration tests. This `EDProducer`  will clone all the event products declared in its configuration, using either the plugin-based NGT trivial serialisation, or the products' ROOT dictionaries.


#### PR tests:

The new unit tests pass.

fwyzard committed:
Introduce ngt::AnyBuffer

ngt::AnyBuffer behaves like std::any, with two differences: it can only be used
with trivially copyable types, and provides access to the underlying memory
buffer to allow memcpy-ing its content.

fwyzard committed:
Introduce ngt::MemoryCopyTraits<T>

The struct ngt::MemoryCopyTraits<T> should be specialised for any type that can
be safely memcpy-ed.

ghyls committed:
Implement Catch2 tests for ngt::MemoryCopyTraits

ghyls committed:
Introduce a plugin-based trivial serialiser

Co-authored-by: Andrea Bocci <andrea.bocci@cern.ch>

ghyls committed:
Introduce ngt::GenericCloner

This EDProducer will clone all the event products declared in its configuration,
using either the plugin-based NGT trivial serialisation, or the products' ROOT
dictionaries.

Co-authored-by: Andrea Bocci <andrea.bocci@cern.ch>

fwyzard committed:
Implement the MemoryCopyTraits for PortableHostCollection and PortableHostObject

Co-authored-by: Mario Gonzalez <mario.gonzalez.carpintero@cern.ch>

ghyls committed:
Implement serialisation plugins for SoA products

Add a test for the serialisation using the GenericCloner.

Co-authored-by: Andrea Bocci <andrea.bocci@cern.ch>

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"0494ef8d4044b1073339041d522c2998a7ff2245":{"files":["HeterogeneousCore/SerialisationCore/test/BuildFile.xml","HeterogeneousCore/SerialisationCore/test/test_catch2_MemoryCopyTraits.cc","HeterogeneousCore/SerialisationCore/test/test_catch2_main.cc"],"squashed":false,"time":1764102486},"186c3893e9cab6695b651402392ced21f1dcda81":{"files":["DataFormats/SiPixelClusterSoA/BuildFile.xml"],"squashed":false,"time":1764086373},"244349d7963ea1f247b337cae9e8ca5bab07527e":{"files":["FWCore/Framework/interface/WrapperBaseHandle.h","FWCore/Framework/interface/WrapperBaseOrphanHandle.h","FWCore/Framework/src/WrapperBaseHandle.cc"],"squashed":false,"time":1764102476},"5ed295cbdaee059a65b3573470ed0837c08f9fda":{"files":["HeterogeneousCore/SerialisationCore/README.md","HeterogeneousCore/SerialisationCore/interface/MemoryCopyTraits.h"],"squashed":false,"time":1764102483},"63bc511482c767a961010ac59a3aec0b9bf1acb2":{"files":["DataFormats/BeamSpot/plugins/BuildFile.xml","DataFormats/BeamSpot/plugins/serialisation.cc","DataFormats/EcalDigi/plugins/BuildFile.xml","DataFormats/EcalDigi/plugins/serialisation.cc","DataFormats/EcalRecHit/plugins/BuildFile.xml","DataFormats/EcalRecHit/plugins/serialisation.cc","DataFormats/HGCalDigi/plugins/BuildFile.xml","DataFormats/HGCalDigi/plugins/serialisation.cc","DataFormats/HGCalReco/plugins/BuildFile.xml","DataFormats/HGCalReco/plugins/serialisation.cc","DataFormats/HcalDigi/plugins/BuildFile.xml","DataFormats/HcalDigi/plugins/serialisation.cc","DataFormats/HcalRecHit/plugins/BuildFile.xml","DataFormats/HcalRecHit/plugins/serialisation.cc","DataFormats/ParticleFlowReco/plugins/BuildFile.xml","DataFormats/ParticleFlowReco/plugins/serialisation.cc","DataFormats/SiPixelClusterSoA/interface/SiPixelClustersHost.h","DataFormats/SiPixelClusterSoA/plugins/BuildFile.xml","DataFormats/SiPixelClusterSoA/plugins/serialisation.cc","DataFormats/SiPixelDigiSoA/interface/SiPixelDigiErrorsHost.h","DataFormats/SiPixelDigiSoA/interface/SiPixelDigisHost.h","DataFormats/SiPixelDigiSoA/plugins/BuildFile.xml","DataFormats/SiPixelDigiSoA/plugins/serialisation.cc","DataFormats/TrackSoA/plugins/BuildFile.xml","DataFormats/TrackSoA/plugins/serialisation.cc","DataFormats/TrackingRecHitSoA/interface/TrackingRecHitsHost.h","DataFormats/TrackingRecHitSoA/plugins/BuildFile.xml","DataFormats/TrackingRecHitSoA/plugins/serialisation.cc","DataFormats/VertexSoA/plugins/BuildFile.xml","DataFormats/VertexSoA/plugins/serialisation.cc"],"squashed":false,"time":1764102498},"689675f36b747d8d9c6285af7d715b0c63a0ab78":{"files":["HeterogeneousCore/SerialisationCore/README.md","HeterogeneousCore/SerialisationCore/plugins/BuildFile.xml","HeterogeneousCore/SerialisationCore/plugins/GenericCloner.cc","HeterogeneousCore/SerialisationCore/test/BuildFile.xml","HeterogeneousCore/SerialisationCore/test/testGenericCloner_cfg.py"],"squashed":false,"time":1764102491},"bf52079f5447f563ae47626132c623d82e3e7806":{"files":["HeterogeneousCore/SerialisationCore/README.md","HeterogeneousCore/SerialisationCore/interface/AnyBuffer.h"],"squashed":false,"time":1764102481},"d85a4733b41d77b7320659f8dc30e245058bf3d4":{"files":["HeterogeneousCore/SerialisationCore/BuildFile.xml","HeterogeneousCore/SerialisationCore/interface/Reader.h","HeterogeneousCore/SerialisationCore/interface/ReaderBase.h","HeterogeneousCore/SerialisationCore/interface/Serialiser.h","HeterogeneousCore/SerialisationCore/interface/SerialiserBase.h","HeterogeneousCore/SerialisationCore/interface/SerialiserFactory.h","HeterogeneousCore/SerialisationCore/interface/Writer.h","HeterogeneousCore/SerialisationCore/interface/WriterBase.h","HeterogeneousCore/SerialisationCore/plugins/BuildFile.xml","HeterogeneousCore/SerialisationCore/plugins/serialisation.cc","HeterogeneousCore/SerialisationCore/src/SerialiserFactory.cc","HeterogeneousCore/SerialisationCore/test/test_catch2_SerialisationPluginFactory.cc"],"squashed":false,"time":1764102488},"ec25c18d4003894377508e07727f3de94e4fcd70":{"files":["DataFormats/Portable/BuildFile.xml","DataFormats/Portable/interface/PortableCollection.h","DataFormats/Portable/interface/PortableCollectionCommon.h","DataFormats/Portable/interface/PortableDeviceCollection.h","DataFormats/Portable/interface/PortableHostCollection.h","DataFormats/Portable/interface/PortableHostObject.h","DataFormats/Portable/interface/PortableObject.h"],"squashed":false,"time":1764102493},"f15c34b5f46d35cee0ddc64ef899d3265edff768":{"files":["FWCore/Integration/plugins/BuildFile.xml","FWCore/Integration/plugins/WrapperBaseProducer.cc","FWCore/Integration/test/BuildFile.xml","FWCore/Integration/test/testWrapperBaseProducer_cfg.py"],"squashed":false,"time":1764102478},"f81f92087e177b9e21b75b89466178276cadc990":{"files":["DataFormats/PortableTestObjects/plugins/BuildFile.xml","DataFormats/PortableTestObjects/plugins/serialisation.cc","HeterogeneousCore/SerialisationCore/test/BuildFile.xml","HeterogeneousCore/SerialisationCore/test/testGenericClonerHost_cfg.py"],"squashed":false,"time":1764102496}},"emoji":{},"last_seen_sha":"63bc511482c767a961010ac59a3aec0b9bf1acb2","signatures":{}} -->

cmsbuild commented:
A new Pull Request was created by  @fwyzard for master.

It involves the following packages:

- DataFormats/BeamSpot (**reconstruction**)
- DataFormats/EcalDigi (**simulation**)
- DataFormats/EcalRecHit (**reconstruction**)
- DataFormats/HGCalDigi (**simulation**)
- DataFormats/HGCalReco (**reconstruction**)
- DataFormats/HcalDigi (**simulation**)
- DataFormats/HcalRecHit (**reconstruction**)
- DataFormats/ParticleFlowReco (**reconstruction**)
- DataFormats/Portable (**heterogeneous**)
- DataFormats/PortableTestObjects (**heterogeneous**)
- DataFormats/SiPixelClusterSoA (**heterogeneous, reconstruction**)
- DataFormats/SiPixelDigiSoA (**heterogeneous, reconstruction**)
- DataFormats/TrackSoA (**heterogeneous, reconstruction**)
- DataFormats/TrackingRecHitSoA (**heterogeneous, reconstruction**)
- DataFormats/VertexSoA (**heterogeneous, reconstruction**)
- FWCore/Framework (**core**)
- FWCore/Integration (**core**)
- HeterogeneousCore/SerialisationCore (****)


The following packages do not have a category, yet:

HeterogeneousCore/SerialisationCore
Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories_map.py to assign category

@Dr15Jones, @civanch, @cmsbuild, @fwyzard, @jfernan2, @kpedro88, @makortel, @mandrenguyen, @mdhildreth, @smuzaffar, @srimanob can you please review it and eventually sign? Thanks.
@IzaakWN, @ReyerBand, @VinInn, @VourMa, @abdoulline, @argiro, @bsunanda, @dkotlins, @elusian, @ferencek, @gpetruc, @hatakeyamak, @lgray, @makortel, @mariadalfonso, @missirol, @mmasciov, @mmusich, @mroguljic, @mtosi, @pfs, @rchatter, @rovere, @thomreis, @tsusa, @wang0jin, @wddgit this is something you requested to watch as well.
@ftenchini, @mandrenguyen, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

fwyzard commented on HeterogeneousCore/SerialisationCore/README.md:
Should this package be simply `HeterogeneousCore/Serialisation` ?
Or something more explicit ?

fwyzard commented:
@ghyls FYI

cmsbuild commented:
+1

**Size**: This PR adds an extra 80KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7ef6e7/49690/summary.html
**COMMIT**: 5c7cf7aaf156ca2b97d479a06cf6b1a8c5117a4b
**CMSSW**: CMSSW_16_0_X_2025-11-25-2300/el8_amd64_gcc13
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/49475/49690/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @fwyzard cms-sw/cmssw#49449
 - @fwyzard cms-sw/cmssw#49110
 - @smuzaffar cms-sw/cmssw#49472
 - @fwyzard cms-sw/cmssw#49471
 - @bsunanda cms-sw/cmssw#49478

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7ef6e7/49690/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7ef6e7/49690/git-merge-result


## Comparison Summary



**Summary**:
 - You potentially [added 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-11-25-2300+7ef6e7/72057/validateJR/logRootQA-events.log) to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 11 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-11-25-2300+7ef6e7/72057/validateJR.html)
 - Reco comparison had  4 failed jobs
 - DQMHistoTests: Total files compared: 55
 - DQMHistoTests: Total histograms compared: 4502606
 - DQMHistoTests: [Total failures: 21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-11-25-2300+7ef6e7/72057/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4502565
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 54 files compared)
 - Checked 235 log files, 208 edm output root files, 55 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-11-25-2300+7ef6e7/72057/triggerResults): no differences found

fwyzard commented:
Rebased on top of CMSSW_16_0_0_pre3.

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49475/47013

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Portable/BuildFile.xml  modified in PR(s): #32632, #49233
  - File  DataFormats/Portable/interface/PortableCollection.h  modified in PR(s): #32632, #49233
  - File  DataFormats/Portable/interface/PortableCollectionCommon.h  modified in PR(s): #49072
  - File  DataFormats/Portable/interface/PortableDeviceCollection.h  modified in PR(s): #49072
  - File  DataFormats/Portable/interface/PortableHostCollection.h  modified in PR(s): #32632, #49072, #49152, #49233
  - File  DataFormats/Portable/interface/PortableHostObject.h  modified in PR(s): #32632, #49152, #49233, #49293
  - File  DataFormats/PortableTestObjects/plugins/BuildFile.xml  modified in PR(s): #49152
  - File  HeterogeneousCore/SerialisationCore/README.md  modified in PR(s): #49235
  - File  HeterogeneousCore/SerialisationCore/interface/AnyBuffer.h  modified in PR(s): #49235

cmsbuild commented:
Pull request #49475 was updated.

cmsbuild commented:
+1

**Size**: This PR adds an extra 84KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7ef6e7/49735/summary.html
**COMMIT**: b1152115bf67197f424db5e3635f80040423ed26
**CMSSW**: CMSSW_16_0_X_2025-11-28-2300/el8_amd64_gcc13
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/49475/49735/install.sh` to create a dev area with all the needed externals and cmssw changes.

**DAS Queries**: The DAS query tests failed, see the summary page for details.


## Comparison Summary



**Summary**:
 - You potentially [removed 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-11-28-2300+7ef6e7/72097/validateJR/logRootQA-events.log) from the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 3 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-11-28-2300+7ef6e7/72097/validateJR.html)
 - Reco comparison had  4 failed jobs
 - DQMHistoTests: Total files compared: 55
 - DQMHistoTests: Total histograms compared: 4502606
 - DQMHistoTests: [Total failures: 67](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-11-28-2300+7ef6e7/72097/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4502519
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 54 files compared)
 - Checked 235 log files, 208 edm output root files, 55 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-11-28-2300+7ef6e7/72097/triggerResults): no differences found

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49475/47015

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Portable/BuildFile.xml  modified in PR(s): #32632, #49233
  - File  DataFormats/Portable/interface/PortableCollection.h  modified in PR(s): #32632, #49233
  - File  DataFormats/Portable/interface/PortableCollectionCommon.h  modified in PR(s): #49072
  - File  DataFormats/Portable/interface/PortableDeviceCollection.h  modified in PR(s): #49072
  - File  DataFormats/Portable/interface/PortableHostCollection.h  modified in PR(s): #32632, #49072, #49152, #49233
  - File  DataFormats/Portable/interface/PortableHostObject.h  modified in PR(s): #32632, #49152, #49233, #49293
  - File  DataFormats/PortableTestObjects/plugins/BuildFile.xml  modified in PR(s): #49152
  - File  HeterogeneousCore/SerialisationCore/README.md  modified in PR(s): #49235
  - File  HeterogeneousCore/SerialisationCore/interface/AnyBuffer.h  modified in PR(s): #49235

cmsbuild commented:
Pull request #49475 was updated.

fwyzard commented:
To do:
  - [ ] move `HeterogeneousCore/SerialisationCore/interface/MemoryCopyTraits.h` to `DataFormats/TrivialSerialisation` to remove the dependency of the other `DataFormats` on ``HeterogeneousCore/SerialisationCore` (which depends on `FWCore`)
  - [ ] rename `HeterogeneousCore/SerialisationCore` to `HeterogeneousCore/TrivialSerialisation`
  - [ ] rename 'serialisation.cc' to 'trivial_serialisation.cc'

fwyzard commented:
Rebased to the current `master`, and updated the development following the discussion at the Core Software meeting.

In particular, `DataFormats/Portable` now depends only on `DataFormats/Common` and `DataFormats/TrivialSerialisation`, and the latter does not have any dependencies.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49475/47193

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/BeamSpot/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/EcalDigi/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/EcalRecHit/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/HGCalDigi/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/HGCalReco/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/HcalDigi/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/HcalRecHit/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/ParticleFlowReco/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/ParticleFlowReco/src/alpaka/classes_cuda.h  modified in PR(s): #32632
  - File  DataFormats/ParticleFlowReco/src/alpaka/classes_rocm.h  modified in PR(s): #32632
  - File  DataFormats/ParticleFlowReco/src/classes_serial.cc  modified in PR(s): #32632
  - File  DataFormats/ParticleFlowReco/src/classes_serial.h  modified in PR(s): #32632
  - File  DataFormats/Portable/BuildFile.xml  modified in PR(s): #32632, #49233
  - File  DataFormats/Portable/interface/PortableCollection.h  modified in PR(s): #32632, #49233
  - File  DataFormats/Portable/interface/PortableCollectionCommon.h  modified in PR(s): #32632, #49072
  - File  DataFormats/Portable/interface/PortableDeviceCollection.h  modified in PR(s): #32632, #49072
  - File  DataFormats/Portable/interface/PortableHostCollection.h  modified in PR(s): #32632, #49072, #49152, #49233
  - File  DataFormats/Portable/interface/PortableHostObject.h  modified in PR(s): #32632, #49152, #49233
  - File  DataFormats/Portable/interface/PortableObject.h  modified in PR(s): #32632
  - File  DataFormats/PortableTestObjects/plugins/BuildFile.xml  modified in PR(s): #32632, #49152
  - File  DataFormats/SiPixelClusterSoA/interface/SiPixelClustersHost.h  modified in PR(s): #32632
  - File  DataFormats/SiPixelClusterSoA/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/SiPixelDigiSoA/interface/SiPixelDigiErrorsHost.h  modified in PR(s): #32632
  - File  DataFormats/SiPixelDigiSoA/interface/SiPixelDigisHost.h  modified in PR(s): #32632
  - File  DataFormats/SiPixelDigiSoA/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/TrackSoA/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/TrackingRecHitSoA/interface/TrackingRecHitsHost.h  modified in PR(s): #32632
  - File  DataFormats/TrackingRecHitSoA/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/VertexSoA/plugins/BuildFile.xml  modified in PR(s): #32632

fwyzard commented:
The new packages are added by https://github.com/cms-sw/cms-bot/pull/2638 .

makortel commented on DataFormats/TrivialSerialisation/interface/MemoryCopyTraits.h:
```diff
@@ -0,0 +1,145 @@
+  concept HasTrivialCopyFinalize = requires(T& object) { ngt::MemoryCopyTraits<T>::finalize(object); };
+
+  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+  // Specialisations for various types
```
Do these specializations need to be visible to everyone who `#include`s this header, or only to the plugins where the `DEFINE_TRIVIAL_SERIALISER_PLUGIN` is used to define the plugins for the specialized types?

fwyzard replied:
In principle the `MemoryCopyTraits` are more generic than the trivial serialisation method - for example, they should provide enough information to reimplement the host/device copies, if we wanted to.

For this reason I would prefer not to move the specialisation to the plugin that uses them.

I can move them to separate headers in the same package, _e.g._ `DataFormats/TrivialSerialisation/interface/arithmetic.h`, `DataFormats/TrivialSerialisation/interface/vector.h`, _etc_ if you think that would be clearer.
but I'm not sure if that would be better ?

makortel commented on HeterogeneousCore/TrivialSerialisation/test/test_catch2_main.cc:
```diff
@@ -0,0 +1,2 @@
+#define CATCH_CONFIG_MAIN
```
I see both this file and `test_catch2_SerialisationPluginFactory.cc` define `CATCH_CONFIG_MAIN`. Do you want to keep this file (and remove the definition from the other`, or remove this file?

fwyzard replied:
Thanks for the catch (🤣), it's a leftover from shuffling the tests around. I fixed `test_catch2_SerialisationPluginFactory.cc`.

cmsbuild commented:
+1

**Size**: This PR adds an extra 88KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7ef6e7/50155/summary.html
**COMMIT**: 10a0353895e02b74117969038598279ca9440319
**CMSSW**: CMSSW_16_0_X_2025-12-16-1100/el8_amd64_gcc13
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/49475/50155/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 3 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-12-16-1100+7ef6e7/72407/validateJR/logRootQA-events.log) from the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 4 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-12-16-1100+7ef6e7/72407/validateJR.html)
 - Reco comparison had  4 failed jobs
 - DQMHistoTests: Total files compared: 55
 - DQMHistoTests: Total histograms compared: 4513634
 - DQMHistoTests: [Total failures: 70](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-12-16-1100+7ef6e7/72407/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4513544
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 54 files compared)
 - Checked 235 log files, 208 edm output root files, 55 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_0_X_2025-12-16-1100+7ef6e7/72407/triggerResults): no differences found

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49475/47194

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/BeamSpot/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/EcalDigi/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/EcalRecHit/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/HGCalDigi/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/HGCalReco/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/HcalDigi/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/HcalRecHit/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/ParticleFlowReco/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/ParticleFlowReco/src/alpaka/classes_cuda.h  modified in PR(s): #32632
  - File  DataFormats/ParticleFlowReco/src/alpaka/classes_rocm.h  modified in PR(s): #32632
  - File  DataFormats/ParticleFlowReco/src/classes_serial.cc  modified in PR(s): #32632
  - File  DataFormats/ParticleFlowReco/src/classes_serial.h  modified in PR(s): #32632
  - File  DataFormats/Portable/BuildFile.xml  modified in PR(s): #32632, #49233
  - File  DataFormats/Portable/interface/PortableCollection.h  modified in PR(s): #32632, #49233
  - File  DataFormats/Portable/interface/PortableCollectionCommon.h  modified in PR(s): #32632, #49072
  - File  DataFormats/Portable/interface/PortableDeviceCollection.h  modified in PR(s): #32632, #49072
  - File  DataFormats/Portable/interface/PortableHostCollection.h  modified in PR(s): #32632, #49072, #49152, #49233
  - File  DataFormats/Portable/interface/PortableHostObject.h  modified in PR(s): #32632, #49152, #49233
  - File  DataFormats/Portable/interface/PortableObject.h  modified in PR(s): #32632
  - File  DataFormats/PortableTestObjects/plugins/BuildFile.xml  modified in PR(s): #32632, #49152
  - File  DataFormats/SiPixelClusterSoA/interface/SiPixelClustersHost.h  modified in PR(s): #32632
  - File  DataFormats/SiPixelClusterSoA/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/SiPixelDigiSoA/interface/SiPixelDigiErrorsHost.h  modified in PR(s): #32632
  - File  DataFormats/SiPixelDigiSoA/interface/SiPixelDigisHost.h  modified in PR(s): #32632
  - File  DataFormats/SiPixelDigiSoA/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/TrackSoA/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/TrackingRecHitSoA/interface/TrackingRecHitsHost.h  modified in PR(s): #32632
  - File  DataFormats/TrackingRecHitSoA/plugins/BuildFile.xml  modified in PR(s): #32632
  - File  DataFormats/VertexSoA/plugins/BuildFile.xml  modified in PR(s): #32632

makortel commented:
Comparison differences are related to https://github.com/cms-sw/cmssw/issues/47071

kpedro88 commented:
+simulation
for minor additions in Digi data format packages

makortel commented:
> I can add only `DataFormats/TrivialSerialisation/interface/MemoryCopyTraits.h` in to [fwlite_build_set](https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_16_0_X/master/fwlite_build_set.file) to fix the fwlite builds . Let me know if this is acceptable

I'd add the entire `DataFormats/TrivialSerialisation` package. As @fwyzard wrote, that package does not have any dependencies.
